common/geth/client.go

```diff
@@ -26,253 +16,54 @@ var (
-	opts.Value = value
 
-	contract := c.Contracts[*tx.To()]
-	// if the contract has not been cached
-	if contract == nil {
-		// create a dummy bound contract tied to the `to` address of the transaction
-		contract = bind.NewBoundContract(*tx.To(), abi.ABI{}, c.Client, c.Client, c.Client)
-		// cache the contract for later use
-		c.Contracts[*tx.To()] = contract
+	if len(config.PrivateKeyStringBackup) != 0 && len(config.PrivateKeyStringBackup) != len(config.RPCURLBackup) {
+		return nil, fmt.Errorf("inconsistent number of %v private keys and %v url", len(config.PrivateKeyStringBackup), len(config.RPCURLBackup))
 	}
-	return contract.RawTransact(opts, tx.Data())
-}
 
-// EstimateGasPriceAndLimitAndSendTx sends and returns a transaction receipt.
-//
-// Note: tx must be a to a contract, not an EOA
-func (c *EthClient) EstimateGasPriceAndLimitAndSendTx(
-	ctx context.Context,
-	tx *types.Transaction,
-	tag string,
-	value *big.Int,
-) (*types.Receipt, error) {
-	gasTipCap, gasFeeCap, err := c.GetLatestGasCaps(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("EstimateGasPriceAndLimitAndSendTx: failed to get gas price for txn (%s): %w", tag, err)
-	}
-
-	tx, err = c.UpdateGas(ctx, tx, value, gasTipCap, gasFeeCap)
-	if err != nil {
-		return nil, fmt.Errorf("EstimateGasPriceAndLimitAndSendTx: failed to update gas for txn (%s): %w", tag, err)
-	}
+	numBackup := len(config.RPCURLBackup)
+	backups := make([]*EthClientBase, numBackup)
+	for i := 0; i < numBackup; i++ {
+		// overwrite the default
+		config.RPCURL = config.RPCURLBackup[i]
+		if len(config.PrivateKeyStringBackup) != 0 {
+			config.PrivateKeyString = config.PrivateKeyStringBackup[i]
+		}
 
-	err = c.SendTransaction(ctx, tx)
-	if err != nil {
-		return nil, fmt.Errorf("EstimateGasPriceAndLimitAndSendTx: failed to send txn (%s): %w", tag, err)
+		backups[i], err = NewClientInstance(config, logger, controller)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	receipt, err := c.EnsureTransactionEvaled(
-		ctx,
-		tx,
-		tag,
-	)
-	if err != nil {
-		return nil, err
-	}
+	logger.Info("Maintain", 1+numBackup, "Eth Client Instances")
 
-	return receipt, err
+	return &EthClient{
+		EthClientBase:   primary,
+		BackupInstances: backups,
+		Controller:      controller,
+	}, nil
 }
 
-// EnsureTransactionEvaled waits for tx to be mined on the blockchain and returns the receipt.
-// If the context times out but the receipt is available, it returns both receipt and error, noting that the transaction is confirmed but has not accumulated the required number of confirmations.
-func (c *EthClient) EnsureTransactionEvaled(ctx context.Context, tx *types.Transaction, tag string) (*types.Receipt, error) {
-	receipt, err := c.waitMined(ctx, []*types.Transaction{tx})
-	if err != nil {
-		return receipt, fmt.Errorf("EnsureTransactionEvaled: failed to wait for transaction (%s) to mine: %w", tag, err)
+func (c *EthClient) GetEthClientInstance() dacommon.EthClient {
+	isPrimary, index := c.Controller.GetClientIndex()
```

Comment:
Since primary and backups are put together for round robin, it seems no need to distinguish a primary, and they can just be treated as a pool of providers